### PR TITLE
State diff model

### LINKED ISSUE #313: StateDiff model
In order to fulfil the snapshot requirements, we need a state diff database model.

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #313 

Still need to handle endianness (will be cared of when endian related file will be created)

# Test Plan:
Existing and added tests

### CODE REVIEW

File: consensus/snapshot/tests/test_state_diff.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include <unity/unity.h>
+
+#include "common/helpers/digest.h"
+#include "consensus/snapshot/snapshot.h"
+#include "consensus/snapshot/state_diff.h"
+
+// TODO Remove after "Common definitions #329" is merged
+#define HASH_LENGTH 243
+
+void test_serialize_deserialize_range() {
```
Comment (tsvisabo) [inline consensus/snapshot/tests/test_state_diff.c]: i think `test_milestone_state_diff` make this test redundant

**Review by tsvisabo — DISMISSED** [review]

File: common/storage/sql/sqlite3/storage.c [hunk 2]
```diff
@@ -639,3 +639,91 @@ retcode_t iota_stor_milestone_exist(const connection_t* const conn,
...
+              "Failed in binding, sqlite3 code is: %" PRIu64 "\n", rc);
+    ret = RC_SQLITE3_FAILED_BINDING;
+    goto done;
+  }
+
+  if ((ret = execute_statement_store(sqlite_statement))) {
+    goto done;
+  }
+
+done:
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c:661]: This is missing a call to `sqlite3_finalize`: https://www.sqlite.org/c3ref/finalize.html

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/storage/sql/sqlite3/storage.c (hunk 2 above)
Comment (thibault-martinez) [reply common/storage/sql/sqlite3/storage.c:661]: It's done in `execute_statement_store`

Comment (thibault-martinez) [reply common/storage/sql/sqlite3/storage.c:661]: I'm doing a storage refactoring in another branch rn.

Comment (thibault-martinez) [reply common/storage/sql/sqlite3/storage.c:661]: Fixed by #398

Comment (th0br0) [reply common/storage/sql/sqlite3/storage.c:661]: Thanks :)

File: common/storage/sql/sqlite3/storage.c [hunk 3]
```diff
@@ -621,3 +621,83 @@ retcode_t iota_stor_milestone_exist(connection_t const* const conn,
...
+  }
+
+  rc = sqlite3_step(sqlite_statement);
+  if (rc == SQLITE_ROW) {
+    bytes = (byte_t*)sqlite3_column_blob(sqlite_statement, 0);
+    size = sqlite3_column_bytes(sqlite_statement, 0);
+    if ((ret = iota_state_diff_deserialize(bytes, size, diff)) != RC_OK) {
+      goto done;
+    }
+  } else if (rc != SQLITE_OK && rc != SQLITE_DONE) {
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: this should probably be `||`? currently this will always be false

**Review by th0br0 — APPROVED** [review]

**Pull request merged**